### PR TITLE
Incipient snapshot mechanism

### PR DESCRIPTION
**Pull request by miloszm**
Incipient snapshot mechanism - contains API for uncompressed and compressed memory snapshots including taking the snapshots as well as deploying modules off the snapshots.
Relevant test is contained in boxen.rs.

### CODE REVIEW

File: hatchery/tests/boxen.rs [hunk 1]
```diff
@@ -54,3 +54,40 @@ pub fn box_set_store_restore_get() -> Result<(), Error> {
...
+    let value: Option<i16> = world.query(id, "get", ())?;
+    assert_eq!(value, Some(TRANSACT_VALUE));
+    let id = world.deploy_from_compressed_snapshot(
+        module_bytecode!("box"),
+        snapshot1.id(),
+        snapshot2_compressed.id(),
+    )?;
+    let value: Option<i16> = world.query(id, "get", ())?;
+    assert_eq!(value, Some(SNAPSHOT2_COMPRESSED_VALUE));
+    let id = world.deploy_from_uncompressed_snapshot(
```
Comment (krl) [inline hatchery/tests/boxen.rs:86]: This is way too much API surface area, the user should not have to worry about different kinds of snapshots. Either we have compression active at all times, or we don't use it.

File: hatchery/tests/stacked.rs [hunk 2]
```diff
@@ -56,3 +58,32 @@ pub fn multi_push_pop() -> Result<(), Error> {
...
+        first_id = first_world.deploy(module_bytecode!("stack"))?;
+        for i in 0..N {
+            first_world.transact(first_id, "push", i)?;
+            let len: i32 = first_world.query(first_id, "len", ())?;
+            assert_eq!(len, i + 1);
+        }
+        first_world.storage_path().clone_into(&mut storage_path);
+    }
+    let mut second_world = World::new(storage_path);
+    let second_id = second_world.deploy(module_bytecode!("stack"))?;
```
Comment (krl) [inline hatchery/tests/stacked.rs:77]: Persisted worlds should include already deployed snapshots, and not have to re-deploy them.

**Review by krl — CHANGES_REQUESTED** [review]
Just some high-level observations and change requests.

**Comment by miloszm** [comment]
superseded by PR 37

**Pull request closed without merging**